### PR TITLE
Replace Builder class with elasticsearch_dsl

### PR DESCRIPTION
**Pull request by hmstepanek**
The Builder class is a customized version of elasticsearch-dsl. It is a legacy class that was made back when there was no solution for building elasticsearch queries. Replacing the Builder class with elasticsearch-dsl reduces hidden complexity that previously existed inside the Builder class (for example making the sorters and limiters into their own qualifier classes) and makes the building of the query more explicit and readable. It also shifts us off of an internally maintained wrapper class to an externally maintained library which means less time that devs have to dedicate to maintaining our internal solution.

### CODE REVIEW

File: h/activity/query.py [hunk 1]
```diff
@@ -165,8 +165,8 @@ def load_documents(query):
 @newrelic.agent.function_trace()
 def _execute_search(request, query, page_size):
     search = Search(request, stats=request.stats)
-    search.append_filter(AuthorityFilter(authority=request.authority))
-    search.append_filter(TopLevelAnnotationsFilter())
+    search.append_qualifier(AuthorityFilter(authority=request.authority))
+    search.append_qualifier(TopLevelAnnotationsFilter())
```
Comment (hmstepanek) [inline h/activity/query.py]: Why not use the separate_replies=True feature instead of appending a qualifier?

File: h/search/core.py [hunk 2]
```diff
@@ -34,14 +35,25 @@ class Search(object):
     :type stats: statsd.client.StatsClient
     """
     def __init__(self, request, separate_replies=False, stats=None, _replies_limit=200):
-        self.request = request
-        self.es = request.es
+        self.es_index = request.es.index
```
Comment (hmstepanek) [inline h/search/core.py]: Note we said goodbye to the request.es object.

File: h/search/core.py [hunk 3]
```diff
@@ -34,14 +35,25 @@ class Search(object):
...
+                            query.DeletedFilter(),
+                            query.AuthFilter(request),
+                            query.UriFilter(request),
+                            query.GroupFilter(),
+                            query.GroupAuthFilter(request),
+                            query.UserFilter(),
+                            query.NipsaFilter(request),
+                            query.AnyMatcher(),
+                            query.TagsMatcher(),
+                            query.KeyValueMatcher()]
```
Comment (hmstepanek) [inline h/search/core.py]: This KeyValueMatcher matches any leftovers-before it was hidden in the depth of the Builder class and now it's more obvious what's going on. Although this catchall thing that has to be run always at the end is kinda awkward. Maybe the improvements we talked about to the api will make this go away?

File: tests/h/search/query_test.py [hunk 4]
```diff
@@ -1,50 +1,12 @@
...
-        ("created", "desc", [2, 0, 1]),
-        ("created", "asc", [1, 0, 2]),
-
-        # Default sort order should be descending.
-        ("updated", None, [1, 0, 2]),
-
-        # Default sort field should be "updated".
-        (None, "asc", [2, 0, 1]),
-    ])
-    def test_it_sorts_annotations(self, Annotation, search, sort_key, order, expected_order):
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply tests/h/search/query_test.py]: ^ There are tests in the old_query_tests class that need to be moved into here. Note that I added three more classes that were hidden inside the Builder class before; Sorter, Limiter, and KeyValueMatcher. There are tests for these in the old_tests and in fact I didn't know we needed a KeyValueMatcher until I fixed up the old tests.

File: h/search/core.py [hunk 5]
```diff
@@ -34,14 +35,25 @@ class Search(object):
...
-        self.es = request.es
+        self.es_index = request.es.index
         self.separate_replies = separate_replies
         self.stats = stats
         self._replies_limit = _replies_limit
-
-        self.builder = self._default_querybuilder(request, self.es)
-        self.reply_builder = self._default_querybuilder(request, self.es)
+        # Order matters!!! The KeyValueMatcher must be run last,
+        # after all other qualifiers have popped off the params.
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: We could but I'm not a huge fan of that 1) because it means the clear method doesn't actually clear everything and 2) it hides things inside the run method. I really like having this list here because it exposes the previously hidden complexity and I know that these are all the things it runs. (Minus the RepliesFilter which I'm ok with cause it's specific to that's how we get replies and the separate_replies feature.) The more we can make things explicit the more readable the code is. For example I didn't even know we had a KeyValueMatcher catch-all until I ran the old_search_tests.

File: h/search/core.py [hunk 6]
```diff
@@ -34,14 +35,25 @@ class Search(object):
...
+        self._qualifiers = [query.Sorter(),
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: I'm not set on the name-I just wanted something more generic.

File: h/search/core.py [hunk 7]
```diff
@@ -53,57 +65,74 @@ def run(self, params):
...
 
-    def _search_annotations(self, params):
-        if self.separate_replies:
-            self.builder.append_filter(query.TopLevelAnnotationsFilter())
+    def _search(self, qualifiers, aggregations, params):
+        """
+        Applies the qualifiers, aggregations, and executes the search.
+        """
+        search = elasticsearch_dsl.Search(
+            using="default", index=self.es_index
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: I don't see a lot of difference here, they are both configured elsewhere in the app. The old way adds an obj and passes it around under the request and dsl uses an alias (the 'default' string). I also don't think the Client class is really necessary if we use dsl. I do think the code looks a lot cleaner now. Notice the old way we are passing all the stuff into the Builder classes's run method, which I find distracting. Now it's very explicit what's happening and every line matters (here I'm referring to the source=False, mapping, etc - those thing really don't matter to me when I'm reading this code).

File: h/search/core.py [hunk 8]
```diff
@@ -53,57 +65,74 @@ def run(self, params):
...
+        for qual in qualifiers:
+            search = qual(search, params)
 
         response = None
         with self._instrument():
-            response = self.es.conn.search(index=self.es.index,
-                                           doc_type=self.es.mapping_type,
-                                           _source=False,
-                                           body=self.builder.build(params))
+            response = search.execute()
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: It looks like you can do something like this:
```
s = s.extra(_source=False)
```
which presumably could be another qualifier.

File: h/search/core.py [hunk 9]
```diff
@@ -53,57 +65,74 @@ def run(self, params):
...
+            qualifiers = [query.TopLevelAnnotationsFilter()] + qualifiers
+
+        response = self._search(qualifiers,
+                                self._aggregations,
+                                params)
+
         total = response['hits']['total']
         annotation_ids = [hit['_id'] for hit in response['hits']['hits']]
-        aggregations = self._parse_aggregation_results(response.get('aggregations', None))
+        aggregations = self._parse_aggregation_results(response.aggregations)
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: the response.aggregations = {}

File: h/search/core.py [hunk 10]
```diff
@@ -53,57 +65,74 @@ def run(self, params):
...
         return (total, annotation_ids, aggregations)
 
     def _search_replies(self, annotation_ids):
         if not self.separate_replies:
             return []
 
-        self.reply_builder.append_matcher(query.RepliesMatcher(annotation_ids))
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: ^ yes. I'm not a huge fan of the list data structure that we are using here for this reason.

File: tests/common/fixtures/elasticsearch.py [hunk 11]
```diff
@@ -22,7 +22,7 @@ def es_client():
                                 conflicts='proceed')
 
 
-@pytest.fixture
+@pytest.fixture(scope="session", autouse=True)
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply tests/common/fixtures/elasticsearch.py]: There are only a few other places in the code that reference request.es. Mainly the BatchIndexer and cli indexing commands. And they don't really need the whole Client class, they just need the connection.

I don't think it's difficult to mock the default connection, it's just:
```
@pytest.fixture
def mocked_connection():
    client = Mock()
    client.search.return_value = dummy_response()
    connections.add_connection('mock', client)
    yield client
```
And note that the connection is not called 'default'. I think if we did make use of the alias then we would want to pass in the alias to things that use it. So everywhere where you see request.es being passed in we would pass in alias and probably index as well. 

The nice thing about the Client class as it exists today or request.es is that it fully defines the connection to a particular index on a particular cluster. I think it serves an important purpose, otherwise you are passing around the index and alias everywhere. And it's nice to know in order to make a different connection all you have to do is create another instance of the Client class. We could still take advantage of the aliasing by using elasticsearch dsl to reference the connection in the Client class. At the end of the day I really don't think it matters either way, whether we create the client via elasticsearch_dsl or via elasticsearch and pass it to elasticsearch_dsl; elasticsearch_dsl excepts both an alias or an elasticsearch client connection as an argument.

File: h/activity/query.py (hunk 1 above)
Comment (hmstepanek) [reply h/activity/query.py]: ^ I realized this later last night after I thought about it for a while. I wonder where the seperate_replies feature is used today...I'm not a huge fan of appending extra filters inside the run query method as I think it is deceitful and non explicit. I wonder if it would be better to just have a do_not_search_for_replies boolean instead?

**Comment by hmstepanek** [comment]
I realize you are on holiday so of course we can talk about this when you get back but I find this solution much more readable than the builder class. It's not completely due to using elasticsearch-dsl but I think it is in part since we no longer have to have something that generates a giant dict. For me, it's nice to not have to go dig into the Builder class to see what extra things it's appending to the query, now it's all contained in the Search class. I also think it's much easier to read something like search.filter("match", field="foo") than {'bool': {'should': [{'match': {'url': 'foobar'}}]}}, and also note elasticsearch-dsl takes care of adding the bool and should so if those ever change in the future elasticsearch-dsl will take care of that change without us having to change our code. We already have a layer of abstraction in our code (the Builder class) so I don't think it's entirely accurate to say elasticsearch-dsl is adding a layer of abstraction; it's really just shifting the layer of abstraction from something we maintain to something someone else maintains.  I find you don't always have to start at elasticsearch docs and the dictionary form you need but rather start instead from the elasticsearch-dsl .filter or .agg.  As far as layers of abstraction go, elasticsearch-dsl doesn't have any hidden surprises like our Builder class does which is why I like it. I hear your point about documentation, I find github is a good source of information on elasticsearch-dsl (the author, who is also an elasticsearch dev, is very responsive) but, yes the in-code documentation is lacking. I'm not proposing we use elasticsearch-dsl for everything. It sounds like it was difficult to use it for the mapping but I think in the case of the Builder class it really made the code a lot more readable and explicit.

File: h/search/core.py (hunk 2 above)
Comment (hmstepanek) [reply h/search/core.py]: Can you explain how there would be unintentional divergence? I could see that that would be the case if we were using both es.conn and elasticsearch-dsl connect, but I don’t see how that would be the case if we were using one or the other. They are only configured in one location AFAIK.

File: h/activity/query.py (hunk 1 above)
Comment (hmstepanek) [reply h/activity/query.py]: replies_limit is not used currently-I'm tempted to remove it, `_seperate_replies` is exposed in our api and potentially used there-the question is, is anyone actually using it when they call our api cause otherwise, it too is not used.

Comment (hmstepanek) [reply h/activity/query.py]: there is no mention of what appears to be private _seperate_replies in our api either so very tempted to remove this as well.

File: h/search/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply h/search/core.py]: The `h.search.core.Search` class only ever returns annotation IDs from the `run` method, so I think the `s.extra(_source=False)` step should go in the class directly rather than being in an external class.

As I see it, the only reason of having the qualifiers/modifiers be separate from the `Search` class is to be able to test their effects independently. We don't need to do that with `_source=False`.

File: h/search/query.py [hunk 12]
```diff
@@ -3,127 +3,99 @@
...
-            # that exists in both ES 1 and ES 6.
-            "unmapped_type": "boolean",
-        }
-    }]
+    def __call__(self, search, params):
+        sort_by = params.pop("sort", "updated")
+        # Sorting must be done on non-analyzed fields.
+        if sort_by == "user":
+            sort_by = "user_raw"
+        return [{
```
Comment (robertknight) [inline h/search/query.py]: This function needs to be updated to follow the `__call__` signature of the other modifier classes (`__call__(self, search: elasticsearch_dsl.Search, params: MultiDict) -> elasticsearch_dsl.Search`). Currently it is still returning a list, which breaks chaining.

File: h/search/query.py [hunk 13]
```diff
@@ -3,127 +3,99 @@
...
+                "order": params.pop("order", "desc"),
+
+                # `unmapped_type` causes unknown fields specified as arguments to
+                # `sort` behave as if all documents contained empty values of the
+                # given type. Without this, specifying eg. `sort=foobar` throws
+                # an exception.
+                #
+                # We use the field type `boolean` to assist with migration because
+                # that exists in both ES 1 and ES 6.
+                "unmapped_type": "boolean",
```
Comment (robertknight) [inline h/search/query.py]: Don't change it in this PR, but I note that we could drop this now since we are validating the `sort` parameter.

File: h/search/query.py [hunk 14]
```diff
@@ -251,56 +223,53 @@ class DeletedFilter(object):
...
+        return self._nipsa_filter(search, self.group_service, self.user)
 
-def nipsa_filter(group_service, user=None):
-    """Return an Elasticsearch filter for filtering out NIPSA'd annotations.
+    def _nipsa_filter(self, search, group_service, user=None):
+        """Return an Elasticsearch filter for filtering out NIPSA'd annotations.
 
-    The returned filter is suitable for inserting into an Es query dict.
-    For example::
+        The returned filter is suitable for inserting into an Es query dict.
```
Comment (robertknight) [inline h/search/query.py]: Any reason not to fold `_nipsa_filter` into the `__call__` method above? Also the documentation is out of date (it doesn't return a filter dict any more). I suggest just removing the no-longer-relevant parts.

File: tests/common/fixtures/elasticsearch.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply tests/common/fixtures/elasticsearch.py]: As you noted - the existing `Client` class encapsulates a bunch of useful things (connection, index). I also like the fact that you can easily trace how it gets initialized and set up because it works in a way that is consistent with other aspects of the code too (eg. `request.db`). The elasticsearch_dsl connection depends on a hidden global inside elasticsearch_dsl.

What I would definitely suggest is that to make this PR safer to deploy, we continue to use the same elasticsearch connection but just change the query construction code.

File: h/search/query.py (hunk 13 above)
Comment (hmstepanek) [reply h/search/query.py]: ^ I also noted this but then I thought we need to be careful about this cause we are validating it in the api, but I'm not sure about the other places where we call search? I don't think we use sort in those places/let it slip through but I'll need to double check before removing.

File: h/search/query.py (hunk 12 above)
Comment (hmstepanek) [reply h/search/query.py]: ^I think you reviewed this at an awkward moment in time so sorry about that. Probably shouldn't have pushed that right away. Anyways this was definitely fixed!

File: h/search/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hmstepanek) [reply h/search/core.py]: Were you wanting a test for this? I'm not sure there's a good way to do this as the code is written right now as it's all hidden within the _search method. I suppose we could mock it and make sure source(False) got called or something?

[the remark this answers is not in the document]
Comment (robertknight) [reply h/search/core.py]: If you did want to test this, I think mocking `elasticsearch_dsl.Search` would be the easiest thing to do. There is no logic involved here so I'm not too worried about it.

File: h/search/query.py [hunk 15]
```diff
@@ -251,56 +223,34 @@ class DeletedFilter(object):
...
     """
 
-    def __call__(self, _):
-        return {"bool": {"must_not": {"exists": {"field": "deleted"}}}}
+    def __call__(self, search, _):
+        return search.exclude("exists", field="deleted")
 
 
 class NipsaFilter(object):
+    """Return an Elasticsearch filter for filtering out NIPSA'd annotations."""
```
Comment (robertknight) [inline h/search/query.py]: This documentation is slightly incorrect in that `__call__` doesn't return an Elasticsearch filter. I think you could reword this to say simply "Filter out hidden (NIPSA'd) annotations".

File: tests/common/fixtures/elasticsearch.py (hunk 11 above)
Comment (robertknight) [inline tests/common/fixtures/elasticsearch.py]: I don't think we need the change to `es_connect` here any more do we? If we're decided on sticking with `request.es` then we could simply remove the whole fixture and associated code.

File: tests/h/search/query_test.py [hunk 16]
```diff
@@ -36,174 +48,97 @@ class TestBuilder(object):
         ("32.7", OFFSET_DEFAULT),
         ("9801", OFFSET_MAX),
     ])
-    def test_offset(self, offset, from_):
-        builder = query.Builder(ES_VERSION)
+    def test_offset(self, pyramid_request, offset, from_):
+        limiter = query.Limiter()
+        search = elasticsearch_dsl.Search(
+            using="default", index=pyramid_request.es.index
```
Comment (robertknight) [inline tests/h/search/query_test.py]: I guess the `index` parameter value doesn't actually matter here?

File: tests/h/search/query_test.py [hunk 17]
```diff
@@ -36,174 +48,97 @@ class TestBuilder(object):
...
+        )
 
+        params = {"offset": offset}
         if offset is MISSING:
-            q = builder.build({})
-        else:
-            q = builder.build({"offset": offset})
+            params = {}
+
+        q = limiter(search, params).to_dict()
```
Comment (robertknight) [inline tests/h/search/query_test.py]: 👍 - Good to have a few examples of testing without having to execute an ES query.
